### PR TITLE
[REVIEW] Added default constructor for SerialTrieNode to solve Thrust bug on CentOS7

### PR DESCRIPTION
**Pull request by jrhemstad**
@mike-wendt reported an error in building libcudf on CentOS7 + CUDA10.

I triaged the issue, and it appears to be some failure in the combination of CUDA10 Thrust + CentOS7 libc++. 

I isolated to a more minimal reproducer:

```
// compile with nvcc --std=c++11 ./test.cu
#include <thrust/device_vector.h>
#include <thrust/host_vector.h>

struct foo {
  //foo() = default; // This code will fail to compile on CentOS7 + CUDA10 without a default ctor
  foo(int n) : value{n} { }
  int value;
};

thrust::host_vector<foo> return_vector(){
  thrust::host_vector<foo> h_vec;
  h_vec.push_back(foo{0});
  return h_vec;
}

int main(void){
  thrust::device_vector<foo> d_vec = return_vector();
}
```

It appears that adding a default constructor for `SerialTrieNode` solved the problem.

### CODE REVIEW

**Comment by vuule** [comment]
Any ideas why the constructor is needed?

My only guess is that creating a device_vector from a host_vector involves filling up the new vector with default values, before the real values are copied over.

File: cpp/src/utilities/trie.cuh [hunk 1]
```diff
@@ -33,6 +33,7 @@ struct SerialTrieNode {
 	int16_t children_offset = -1;
 	char character;
 	bool is_leaf;
+    SerialTrieNode() = default; // FIXME This is necessary for a Thrust bug on CentOS7 + CUDA10
```
Comment (vuule) [inline cpp/src/utilities/trie.cuh]: AFAIK calling this constructor will create an object with undefined values. Given how easy it is to overlook the calls to the default constructor, I would like it to behave consistently.

Please initialize *character* and *is_leaf*, either in an init list, or inline. Using *trie_terminating_character* would make sense for default value of *character*, *is_leaf* can be false.

**Comment by jrhemstad** [comment]
> Any ideas why the constructor is needed?
> 
> My only guess is that creating a device_vector from a host_vector involves filling up the new vector with default values, before the real values are copied over.

Unclear. At some point Thrust is trying to default initialize the elements of the vector, but _only_ for the unique combination of CentOS 7 and CUDA 10.

File: cpp/src/utilities/trie.cuh (hunk 1 above)
Comment (jrhemstad) [reply cpp/src/utilities/trie.cuh]: Sure, my hope is that the default can just be deleted once the Thrust bug is resolved, but makes sense to make it safe in the meantime.

I'll add default in-class initializers.

**Pull request merged**